### ISSUE TITLE
While iterating over a DbSet, if an exception occurs, the exception isn't thrown until after all rows have been returned

### ISSUE THREAD

**Issue by jemiller0** [dotnet/efcore#7858]
The following code iterates over a DbSet in a DbContext. The first time through the loop, an exception is thrown. The Bibs DbSet has close to 8 million rows in it, including a CLOB field which make iterating over all rows slow. What's weird about the following is that the exception doesn't happen right away. Somehow even though the exception happens on the first row, all rows are read back from the database and then after all rows have been read, it throws the exception. I don't know that this is actually a bug in EF Core. Maybe it's just how LINQ or iterators work?

```
using (var oc = new OleLibrary.OleEntityFrameworkCoreContext())
{
    foreach (var b in oc.Bibs.AsNoTracking())
    {
        throw new Exception();
    }
}
```

The following are the log statements from the above code. The application took over 7 minutes to run the above code and throw the exception.

```
OleLibrary Verbose: 0 : Compiling query model:
'(from Bib <generated>_1 in DbSet<Bib>
select <generated>_1)
.AsNoTracking()'
    ThreadId=1
    DateTime=2017-03-12T02:00:40.1859241Z
OleLibrary Verbose: 0 : Optimized query model:
'from Bib <generated>_1 in DbSet<Bib>
select <generated>_1'
    ThreadId=1
    DateTime=2017-03-12T02:00:40.2015494Z
OleLibrary Verbose: 0 : TRACKED: False
(QueryContext queryContext) => IEnumerable<Bib> _ShapedQuery(
    queryContext: queryContext,
    shaperCommandContext: SelectExpression:
        SELECT `o`.`bib_id`, `o`.`content`, `o`.`date_created`, `o`.`created_by`, `o`.`fast_add`, `o`.`former_id`, `o`.`unique_id_prefix`, `o`.`date_updated`, `o`.`updated_by`, `o`.`staff_only`, `o`.`status`, `o`.`status_updated_date`, `o`.`status_updated_by`
        FROM `ole_ds_bib_t` AS `o`
    ,
    shaper: UnbufferedEntityShaper<Bib>
)

    ThreadId=1
    DateTime=2017-03-12T02:00:40.2796797Z
OleLibrary Verbose: 0 : Executed DbCommand (18ms) [Parameters=[], CommandType='Text', CommandTimeout='90']
SELECT `o`.`bib_id`, `o`.`content`, `o`.`date_created`, `o`.`created_by`, `o`.`fast_add`, `o`.`former_id`, `o`.`unique_id_prefix`, `o`.`date_updated`, `o`.`updated_by`, `o`.`staff_only`, `o`.`status`, `o`.`status_updated_date`, `o`.`status_updated_by`
FROM `ole_ds_bib_t` AS `o`
    ThreadId=1
    DateTime=2017-03-12T02:00:40.3265690Z
OleLibrary Verbose: 0 : Closing connection to database 'ole2' on server 'localhost'.
    ThreadId=1
    DateTime=2017-03-12T02:07:57.0445858Z
LibraryConsoleApplication Critical: 0 : System.Exception: Exception of type 'System.Exception' was thrown.
   at LibraryConsoleApplication.Program2.Test1() in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 288
   at LibraryConsoleApplication.Program2.Main(String[] args) in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 71
    ThreadId=1
    DateTime=2017-03-12T02:07:57.2144268Z
LibraryConsoleApplication Information: 0 : Ending
    ThreadId=1
    DateTime=2017-03-12T02:07:57.2826521Z
LibraryConsoleApplication Information: 0 : 00:07:26.9578978 elapsed
    ThreadId=1
    DateTime=2017-03-12T02:07:57.2826521Z
```

I'm using Pomelo's EF Core provider with MySqlConnector.

What is even more confusing is if you use Oracle's provider for EF 6. EF 6 with Oracle's provider masks the real exception and throws a TimeoutException. I've been having a lot of issues with timeout problems and Oracle's driver. I only discovered this cause of it today.

```
OleLibrary Verbose: 0 : Opened connection at 3/11/2017 7:57:42 PM -06:00

    ThreadId=1
    DateTime=2017-03-12T01:57:42.2401272Z
OleLibrary Verbose: 0 : SELECT
`Extent1`.`bib_id`,
`Extent1`.`former_id`,
`Extent1`.`fast_add`,
`Extent1`.`staff_only`,
`Extent1`.`created_by`,
`Extent1`.`date_created`,
`Extent1`.`updated_by`,
`Extent1`.`date_updated`,
`Extent1`.`status`,
`Extent1`.`status_updated_by`,
`Extent1`.`status_updated_date`,
`Extent1`.`unique_id_prefix`,
`Extent1`.`content`
FROM `ole_ds_bib_t` AS `Extent1`
    ThreadId=1
    DateTime=2017-03-12T01:57:45.4903052Z
OleLibrary Verbose: 0 :

    ThreadId=1
    DateTime=2017-03-12T01:57:45.4903052Z
OleLibrary Verbose: 0 : -- Executing at 3/11/2017 7:57:45 PM -06:00

    ThreadId=1
    DateTime=2017-03-12T01:57:45.4903052Z
OleLibrary Verbose: 0 : -- Completed in 2 ms with result: EFMySqlDataReader

    ThreadId=1
    DateTime=2017-03-12T01:57:45.5059312Z
OleLibrary Verbose: 0 :

    ThreadId=1
    DateTime=2017-03-12T01:57:45.5059312Z
LibraryConsoleApplication Critical: 0 : System.TimeoutException: Timeout in IO operation
   at MySql.Data.MySqlClient.TimedStream.StopTimer()
   at MySql.Data.MySqlClient.TimedStream.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.BufferedStream.Read(Byte[] array, Int32 offset, Int32 count)
   at MySql.Data.MySqlClient.MySqlStream.ReadFully(Stream stream, Byte[] buffer, Int32 offset, Int32 count)
   at MySql.Data.MySqlClient.MySqlStream.LoadPacket()
   at MySql.Data.MySqlClient.MySqlStream.ReadPacket()
   at MySql.Data.MySqlClient.NativeDriver.FetchDataRow(Int32 statementId, Int32 columns)
   at MySql.Data.MySqlClient.ResultSet.Close()
   at MySql.Data.MySqlClient.MySqlDataReader.NextResult()
   at MySql.Data.MySqlClient.MySqlDataReader.Close()
   at MySql.Data.MySqlClient.MySqlDataReader.Dispose()
   at MySql.Data.Entity.EFMySqlDataReader.Dispose(Boolean disposing)
   at System.Data.Entity.Core.Common.Internal.Materialization.Shaper`1.Finally()
   at System.Data.Entity.Internal.LazyEnumerator`1.Dispose()
   at LibraryConsoleApplication.Program2.Test1() in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 288
   at LibraryConsoleApplication.Program2.Main(String[] args) in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 71
    ThreadId=1
    DateTime=2017-03-12T01:59:16.7806157Z
LibraryConsoleApplication Information: 0 : Ending
    ThreadId=1
    DateTime=2017-03-12T01:59:16.8431207Z
LibraryConsoleApplication Information: 0 : 00:02:00.0319749 elapsed
    ThreadId=1
    DateTime=2017-03-12T01:59:16.8431207Z

```

What I'm wondering is, is there a way to make the exception happen immediately rather than having to wait for it to iterate through all the rows? I'm not sure if this is an EF thing, or, just how iterators and/or LINQ works.

Also, does anyone know if there is a way to disable the timeout that occurs with Oracle's driver or somehow make it so that it throws the actual exception rather than the TimeoutException?

This would be less confusing if I were able to simply call .ToArray() and get back all the results at once. However, the results are large, so, I need to stream the data. It would consume too much memory to read it all into memory.

I tried the query using the SQL Server provider and the Npgsql provider as well. I had to cancel the SQL Server test because I ran into issue #7363 where it looks like it has a memory leak. After memory grew to 23 GB, I cancelled it out. Npgsql threw the exception after 2 minutes and 45 seconds. I'm assuming it must have timeout, but, unlike Oracle's EF 6, it's throwing the correct exception.

```
OleLibrary Verbose: 0 : Compiling query model:
'(from Bib <generated>_1 in DbSet<Bib>
select <generated>_1)
.AsNoTracking()'
    ThreadId=1
    DateTime=2017-03-12T02:40:59.9607280Z
OleLibrary Verbose: 0 : Optimized query model:
'from Bib <generated>_1 in DbSet<Bib>
select <generated>_1'
    ThreadId=1
    DateTime=2017-03-12T02:40:59.9607280Z
OleLibrary Verbose: 0 : TRACKED: False
(QueryContext queryContext) => IEnumerable<Bib> _ShapedQuery(
    queryContext: queryContext,
    shaperCommandContext: SelectExpression:
        SELECT "o"."bib_id", "o"."content", "o"."date_created", "o"."created_by", "o"."fast_add", "o"."former_id", "o"."unique_id_prefix", "o"."date_updated", "o"."updated_by", "o"."staff_only", "o"."status", "o"."status_updated_date", "o"."status_updated_by"
        FROM "ole_ds_bib_t" AS "o"
    ,
    shaper: UnbufferedEntityShaper<Bib>
)

    ThreadId=1
    DateTime=2017-03-12T02:41:00.0609919Z
OleLibrary Verbose: 0 : Opening connection to database 'ole2' on server 'tcp://localhost:5432'.
    ThreadId=1
    DateTime=2017-03-12T02:41:01.2476431Z
OleLibrary Verbose: 0 : Executed DbCommand (68ms) [Parameters=[], CommandType='Text', CommandTimeout='90']
SELECT "o"."bib_id", "o"."content", "o"."date_created", "o"."created_by", "o"."fast_add", "o"."former_id", "o"."unique_id_prefix", "o"."date_updated", "o"."updated_by", "o"."staff_only", "o"."status", "o"."status_updated_date", "o"."status_updated_by"
FROM "ole_ds_bib_t" AS "o"
    ThreadId=1
    DateTime=2017-03-12T02:41:01.5481135Z
OleLibrary Verbose: 0 : Closing connection to database 'ole2' on server 'tcp://localhost:5432'.
    ThreadId=1
    DateTime=2017-03-12T02:43:34.2732386Z
LibraryConsoleApplication Critical: 0 : System.Exception: Exception of type 'System.Exception' was thrown.
   at LibraryConsoleApplication.Program2.Test1() in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 288
   at LibraryConsoleApplication.Program2.Main(String[] args) in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 71
    ThreadId=1
    DateTime=2017-03-12T02:43:34.5050183Z
LibraryConsoleApplication Information: 0 : Ending
    ThreadId=1
    DateTime=2017-03-12T02:43:34.5584070Z
LibraryConsoleApplication Information: 0 : 00:02:45.0764489 elapsed
    ThreadId=1
    DateTime=2017-03-12T02:43:34.5584070Z
```

EF Core version: 1.1.1
Database Provider: Pomelo MySQL, Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Anniversary Update
IDE: Visual Studio 2017

**Comment by EdwardBlair** [comment]
Not an EF issue, I don't think.. Try setting Connectiontimeout to like 10 minutes or something. Or write quicker queries. and do it in batches e.g.

**Comment by jemiller0** [comment]
I just tried a similar test using the following Dapper method with buffered = false to see how it behaves compared to EF.

`public static IEnumerable<dynamic> Query(this IDbConnection cnn, string sql, object param = null, IDbTransaction transaction = null, bool buffered = true, int? commandTimeout = default(int?), CommandType? commandType = default(CommandType?));
`

```
public class OleDapperContext : IDisposable
{
...
    public IEnumerable<dynamic> Query(string sql, object param = null) => Connection.Query(sql, param, Transaction, buffered: false);
...
}

```

```
using (var odc = new OleLibrary.OleDapperContext())
{
    foreach (var b in odc.Query("SELECT * FROM ole_ds_bib_t"))
    {
        throw new Exception();
    }
}

```


SQL Server failed immediately and threw the expected exception.

```
LibraryConsoleApplication Critical: 0 : System.Exception: Exception of type 'System.Exception' was thrown.
   at LibraryConsoleApplication.Program2.Test2() in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 300
   at LibraryConsoleApplication.Program2.Main(String[] args) in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 72
    ThreadId=1
    DateTime=2017-03-12T16:49:18.1373917Z
LibraryConsoleApplication Information: 0 : Ending
    ThreadId=1
    DateTime=2017-03-12T16:49:18.2216706Z
LibraryConsoleApplication Information: 0 : 00:00:00.2917823 elapsed
    ThreadId=1
    DateTime=2017-03-12T16:49:18.2216706Z

```

Npgsql failed immediately, but, masked the original exception. I checked the inner exception and it was null.

```
LibraryConsoleApplication Critical: 0 : Npgsql.PostgresException (0x80004005): 57014: canceling statement due to user request
   at Npgsql.NpgsqlConnector.<DoReadMessage>d__148.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Npgsql.NpgsqlConnector.<ReadMessage>d__147.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Npgsql.NpgsqlConnector.<SkipUntil>d__153.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Npgsql.NpgsqlDataReader.<SkipUntil>d__34.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Npgsql.NpgsqlDataReader.<Consume>d__50.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Npgsql.NpgsqlDataReader.Close(Boolean connectionClosing)
   at Dapper.SqlMapper.<QueryImpl>d__125`1.<>m__Finally1()
   at Dapper.SqlMapper.<QueryImpl>d__125`1.System.IDisposable.Dispose()
   at LibraryConsoleApplication.Program2.Test2() in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 300
   at LibraryConsoleApplication.Program2.Main(String[] args) in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 72
    ThreadId=1
    DateTime=2017-03-12T16:48:12.1982986Z
LibraryConsoleApplication Information: 0 : Ending
    ThreadId=1
    DateTime=2017-03-12T16:48:12.2608591Z
LibraryConsoleApplication Information: 0 : 00:00:01.2543607 elapsed
    ThreadId=1
    DateTime=2017-03-12T16:48:12.2608591Z

```

The MySqlConnector provider (the open source one, not Oracle's) appears to have read through all the results, but, threw the expected exception.

```
LibraryConsoleApplication Critical: 0 : System.Exception: Exception of type 'System.Exception' was thrown.
   at LibraryConsoleApplication.Program2.Test2() in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 300
   at LibraryConsoleApplication.Program2.Main(String[] args) in C:\Users\jemiller\Documents\Visual Studio 2017\Projects\Library\LibraryConsoleApplication\Program2.cs:line 72
    ThreadId=1
    DateTime=2017-03-12T16:57:39.0216674Z
LibraryConsoleApplication Information: 0 : Ending
    ThreadId=1
    DateTime=2017-03-12T16:57:39.0589589Z
LibraryConsoleApplication Information: 0 : 00:07:49.6863599 elapsed
    ThreadId=1
    DateTime=2017-03-12T16:57:39.0589589Z

```

So, it appears that it should be possible to fail out immediately. Although, Dapper returns an IEnumerable<T> instead of an IQueryable<T>. The behavior appears to depend at least in part on the provider used.

If there is a way to make it fail immediately and throw the expected exception in EF Core, I think it would help anyone who is streaming results. 

It took me awhile to figure out what was going on in an app of mine because I was pulling data from one data source and writing to another. I noticed that it looked like it got hung on the write side because it stopped inserting rows. However, using Resource Monitor, I could see that it was till pulling data from the other database. Pretty confusing.

**Comment by ajcvickers** [comment]
@jemiller0 Could you provide a project to reproduce this? In particular, we are not sure what EF provider you are using, or what "OleLibrary" is. On SQL Server, assuming "Bibs" is a normal DbSet, this should not be buffering. But there may be something else going on, which is why a repro project would be very useful.

**Comment by jemiller0** [comment]
@ajcvickers Bibs is a normal DbSet in an EF Core DbContext.  You would need to have a table with a lot of data in it to see the problem. I think you should be able to reproduce the problem with any table that has lots of data in it. I could create a repro, but, the main issue is that you need to have a large table. Otherwise, the code is trivial. Basically, what you see above. Just iterate over the DbSet and throw an exception in the for loop. Then, watch Resource Monitor and you can see that although the exception is thrown the first time through the loop, you can see the network pulling data long after that. I'm wondering if what's happening is that Dispose() is being called on the connection and the connection is reading all data from the server or something before closing the connection.

**Comment by divega** [comment]
@jemiller0 @ajcvickers I wouldn't be too surprised if there are providers (ours or otherwise) that consume all remaining results when disposing the enumerator of the query. It could be just a bug or a requirement of the ADO.NET provider/database. Some additional cleanup may also be needed when the connection is closed if there are open results. 

If this repros using our EF Core providers (e.g. SQL Server and SQLite) we can look into it. If it happens with other providers my recommendation would be to file issues in their respective issue trackers.

**Comment by jemiller0** [comment]
I think the problem will happen with the SQL Server provider. However, I ran into a separate issue #7363 which causes problems testing that (for the table size that I was testing on). In the test that I did with SQL Server I cancelled out once memory grew to 23 GB. As far as I can tell, it was doing the same thing as the MySQL driver. I.e. kept reading data although an exception was throw on the first iteration.

**Comment by smitpatel** [comment]
I tried reproducing this. I created database with 1mil. rows and threw exception on 10th iteration using SqlServer as provider. I got exception in less than a second with approx. 20MB of memory used overall. This seems to be consequence of #7363 , if streaming is not working correctly then it is possible that exception won't be throw immediately (because time would be being spent in fetching all data in memory)

**Comment by jemiller0** [comment]
I think if the table isn't that wide, it might not be that noticeable. The table I'm querying has a CLOB field in it and the size on disk is 31 GB. I only saw the problem in #7363 when using the SQL Server provider.

**Comment by ajcvickers** [comment]
@smitpatel Since you started investigation on this, could you continue? 😁 Can you try with a wide table and see if it repros then. If so, it could be that this and #7363 are dupes. Also, if you can repro it can you check whether it repros with pure ADO.NET/SQL Client code?

**Comment by smitpatel** [comment]
Closing this as dupe #7363 .
If we are not going to stream the results then throwing exception will happen only after buffering has finished.
